src/main/java/org/springframework/data/repository/core/support/RepositoryComposition.java

```diff
@@ -354,21 +356,30 @@ private static RepositoryFragments concat(Stream<RepositoryFragment<?>> left, St
 		/**
 		 * Invoke {@link Method} by resolving the
 		 *
-		 * @param method
+		 * @param invokedMethod invoked method as per invocation on the interface.
+		 * @param methodToCall backend method that is backing the call.
 		 * @param args
 		 * @return
 		 * @throws Throwable
 		 */
-		public Object invoke(Method method, Object[] args) throws Throwable {
+		@Nullable
+		public Object invoke(Method invokedMethod, Method methodToCall, Object[] args) throws Throwable {
 
-			RepositoryFragment<?> fragment = fragmentCache.computeIfAbsent(method, this::findImplementationFragment);
+			RepositoryFragment<?> fragment = fragmentCache.computeIfAbsent(methodToCall, this::findImplementationFragment);
 			Optional<?> optional = fragment.getImplementation();
 
 			if (!optional.isPresent()) {
-				throw new IllegalArgumentException(String.format("No implementation found for method %s", method));
+				throw new IllegalArgumentException(String.format("No implementation found for method %s", methodToCall));
+			}
+
+			ImplementationInvocationMetadata invocationMetadata = invocationMetadataCache.get(invokedMethod);
```

Comment:
`….computeIfAbsent(…)?`

Reply:
This is intended to prevent long-running locks. `ConcurrentHashMap` locks segments while the computation function is being invoked. We had previously bug reports regarding contention and I figured to apply the non-blocking approach directly (i.e. without waiting for a bug report).